index.js

```diff
@@ -1,50 +1,77 @@
-const debug = require('debug')('x-ray:phantom');
-const normalize = require('normalizeurl');
-const Nightmare = require('nightmare');
+const debug = require('debug')('x-ray:puppeteer');
+// const normalize = require('normalizeurl');
+const puppeteer = require('puppeteer');
 
 
 function makeDriver(opts) {
   opts = opts || {};
+  // add default user agent
+  if (!opts.useragent) opts.useragent = 'xray/puppeteer';
 
-	return (ctx, callback) => {
-		debug('going to %s', ctx.url);
-
-    const nightmare = new Nightmare(opts);
-    this.instance = nightmare;
+  return (ctx, callback) => {
+    debug('going to %s', ctx.url);
+    this.instance = puppeteer;
 
     // Setup standard listeners
-    nightmare
-      .useragent(opts.useragent)
-      .on('error', (msg) => debug('client-side javascript error %s', msg))
-      .on('timeout', (timeout) => callback(new Error(timeout)))
-      .on('resourceReceived', (resource) => {
-        if (normalize(resource.url) == normalize(ctx.url)) {
-          debug('got response from %s: %s', resource.url, resource.status);
-          ctx.status = resource.status;
-        };
-      })
-      .on('urlChanged', (url) => {
-        debug('redirect: %s', url);
-        ctx.url = url;
+    // nightmare
+    //   .useragent(opts.useragent)
+    //   .on('error', (msg) => debug('client-side javascript error %s', msg))
+    //   .on('timeout', (timeout) => callback(new Error(timeout)))
+    //   .on('resourceReceived', (resource) => {
+    //     if (normalize(resource.url) == normalize(ctx.url)) {
+    //       debug('got response from %s: %s', resource.url, resource.status);
+    //       ctx.status = resource.status;
+    //     }
+    //   })
+    //   .on('urlChanged', (url) => {
+    //     debug('redirect: %s', url);
+    //     ctx.url = url;
+    //   });
+
+    const autoScroll = async (page) => {
+      await page.evaluate(async () => {
+        await new Promise((resolve, reject) => {
+          var totalHeight = 0;
+          var distance = 100;
+          var timer = setInterval(() => {
+            var scrollHeight = document.body.scrollHeight;
+            window.scrollBy(0, distance);
+            totalHeight += distance;
+
+            if (totalHeight >= scrollHeight) {
+              clearInterval(timer);
+              resolve();
+            }
+          }, 100);
+        });
```

Comment:
This is needed for lazy loading images or when the site needs interaction to load items.
https://blog.cloudlayer.io/puppeteer-waituntil-options/